rust/lance-linalg/src/kmeans.rs

```diff
@@ -187,67 +183,43 @@ pub struct KMeanMembership {
 
-        // New centroids for each cluster
-        let means = stream::iter(0..self.k)
-            .zip(repeat_with(|| {
-                (
-                    self.data.clone(),
-                    cluster_ids.clone(),
-                    self.centroids.clone(),
-                )
-            }))
-            .map(
-                |(cluster, (data, cluster_ids, prev_centroids))| async move {
-                    tokio::task::spawn_blocking(move || {
-                        let mut sum = vec![0.0; dimension];
-                        let data = data.values();
-                        let mut total = 0.0;
-                        // Eager group-by cluster id.
-                        for i in 0..cluster_ids.len() {
-                            if cluster_ids[i] as usize == cluster {
-                                // TODO: use simd ADD
-                                for j in 0..dimension {
-                                    sum[j] += data[i * dimension + j];
-                                }
-                                total += 1.0;
-                            };
-                        }
-                        if total > 0.0 {
-                            let s = Float32Array::from(
-                                sum
-                            );
-                            s.unary_mut(|x| x / total).unwrap()
-                        } else {
-                            warn!("Warning: KMean: cluster {} has no value, does not change centroids.", cluster);
-                            prev_centroids.slice(cluster * dimension, dimension)
-                        }
-                    })
-                    .await
-                },
-            )
-            .buffered(num_cpus::get())
-            .try_collect::<Vec<_>>()
-            .await.map_err(|e| {
-                ArrowError::ComputeError(format!(
-                    "KMeans: failed to compute new centroids: {}",
-                    e
-                ))
-            })?;
-
-        // TODO: concat requires `&[&dyn Array]`. Are there cheaper way to pass Vec<Float32Array> to `concat`?
-        let mut mean_refs: Vec<&dyn Array> = vec![];
-        for m in means.iter() {
-            mean_refs.push(m);
+        let mut cluster_cnts = vec![0_usize; self.k];
+        let mut new_centroids = vec![0.0_f32; self.k * dimension];
+        self.data
+            .values()
+            .chunks_exact(dimension)
+            .zip(self.cluster_ids.iter())
+            .for_each(|(v, c)| {
+                cluster_cnts[*c as usize] += 1;
+                for i in 0..dimension {
+                    new_centroids[*c as usize * dimension + i] += v[i];
+                }
+            });
+        for i in 0..self.k {
+            if cluster_cnts[i] == 0 {
+                warn!("KMeans: cluster {} is empty", i);
+            } else {
+                for j in 0..dimension {
+                    new_centroids[i * dimension + j] /= cluster_cnts[i] as f32;
```

Comment:
Is there any possibility that `cluster_cnts[i]` is zero here?  Maybe if K is larger than the # of data points?

Reply:
Yea. the training algorithm does not work that well to make sure no cluster at the moment.